File: back-bardak/src/main/java/kz/bardak/game/rules/Trump.java
package kz.bardak.game.rules;

import java.util.Objects;

/**
 * Козырь раздачи и вытекающая из него защищённая масть — вместе, потому что порознь они
 * бессмысленны: защищённая масть определяется козырем и меняется вместе с ним
 * (§1.1.1, §1.9).
 *
 * <p>⭐ Базовое допущение дурака «козырь бьёт всё» здесь неверно. Защищённую масть козырь
 * не берёт, поэтому старшинство — отдельная функция {@link #beats(Card, Card)}, а не
 * выражение из двух условий.
 */
public record Trump(Suit suit) {

    /** Защищённая масть по умолчанию — пики. */
    private static final Suit DEFAULT_PROTECTED_SUIT = Suit.SPADES;

    /** Если козырь сам оказался пиками, роль защищённой масти переходит к трефам. */
    private static final Suit FALLBACK_PROTECTED_SUIT = Suit.CLUBS;

    public Trump {
        Objects.requireNonNull(suit, "suit");
    }

    public static Trump of(final Suit suit) {
        return new Trump(suit);
    }

    /**
     * Масть, которую козырь не берёт: пики, а при козырных пиках — трефы (§1.1.1).
     * Защищённая масть есть всегда, и она всегда одна.
     */
    public Suit protectedSuit() {
        return suit == DEFAULT_PROTECTED_SUIT ? FALLBACK_PROTECTED_SUIT : DEFAULT_PROTECTED_SUIT;
    }

    /**
     * Бьёт ли карта защиты карту атаки (§1.1.1, §3.1). Правило применяется исключительно
     * в момент защиты: на атаку, подкидывание и перевод защищённая масть не влияет.
     *
     * @param defence карта, которой пытаются отбиться
     * @param attack  карта на столе, которую пытаются побить
     */
    public boolean beats(final Card defence, final Card attack) {
        Objects.requireNonNull(defence, "defence");
        Objects.requireNonNull(attack, "attack");
        if (defence instanceof JokerCard) {
            return true;
        }
        if (attack instanceof JokerCard) {
            return false;
        }
        final PipCard defencePip = (PipCard) defence;
        final PipCard attackPip = (PipCard) attack;
        if (defencePip.suit() == attackPip.suit()) {
            return defencePip.rank().isHigherThan(attackPip.rank());
        }
        if (attackPip.suit() == protectedSuit()) {
            return false;
        }
        return defencePip.suit() == suit;
    }
}

File: back-go/internal/domain/game/trump.go
package game

// Trump — козырь раздачи и вытекающая из него защищённая масть.
//
// Вместе, потому что порознь они бессмысленны: защищённая масть определяется козырем
// и меняется вместе с ним.
//
// ⭐ Базовое допущение дурака «козырь бьёт всё» здесь НЕВЕРНО. Защищённую масть козырь
// не берёт, поэтому старшинство — отдельная функция Beats, а не выражение из двух условий.
type Trump struct {
	Suit Suit
}

// Защищённая масть по умолчанию — пики; если козырь сам пики, роль переходит к трефам.
const (
	defaultProtectedSuit  = Spades
	fallbackProtectedSuit = Clubs
)

// NewTrump собирает козырь.
func NewTrump(suit Suit) Trump { return Trump{Suit: suit} }

// ProtectedSuit — масть, которую козырь не берёт.
//
// Защищённая масть есть всегда, и она всегда одна.
func (t Trump) ProtectedSuit() Suit {
	if t.Suit == defaultProtectedSuit {
		return fallbackProtectedSuit
	}
	return defaultProtectedSuit
}

// Beats — бьёт ли карта защиты карту атаки.
//
// ⚠️ Правило применяется ИСКЛЮЧИТЕЛЬНО в момент защиты: на атаку, подкидывание и перевод
// защищённая масть не влияет. Смешать это — значит запретить подкидывать пики.
func (t Trump) Beats(defence, attack Card) bool {
	// Джокер кроет что угодно, включая другой джокер.
	if _, ok := defence.(Joker); ok {
		return true
	}
	// Джокер в атаке не берётся ничем, кроме джокера, — а тот отработал выше.
	if _, ok := attack.(Joker); ok {
		return false
	}

	defencePip, okDefence := defence.(Pip)
	attackPip, okAttack := attack.(Pip)
	if !okDefence || !okAttack {
		return false
	}

	// Внутри одной масти решает ранг — козырность роли не играет.
	if defencePip.Suit == attackPip.Suit {
		return defencePip.Rank.IsHigherThan(attackPip.Rank)
	}

	// ⭐ Вот здесь дурак и бардак расходятся: защищённую масть чужой мастью не взять,
	// даже козырем. Отбиться от неё можно только старшей картой той же масти или джокером.
	if attackPip.Suit == t.ProtectedSuit() {
		return false
	}

	return defencePip.Suit == t.Suit
}
